python/ray/data/_internal/execution/interfaces.py

```diff
@@ -0,0 +1,274 @@
+                return len(ready) > 0
+
+            def get_next(self):
+                ready, remaining = ray.wait(self.active_tasks, num_returns=1)
+                self.active_tasks = remaining
+                return ready[0]
+
+    Note that the above operator fully supports both bulk and streaming execution,
+    since `add_input` and `get_next` can be called in any order. In bulk execution,
+    all inputs would be added up-front, but in streaming execution the calls could
+    be interleaved.
+    """
+
+    def __init__(self, name: str, input_dependencies: List["PhysicalOperator"]):
+        self._name = name
+        self._input_dependencies = input_dependencies
+        for x in input_dependencies:
+            assert isinstance(x, PhysicalOperator), x
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def input_dependencies(self) -> List["PhysicalOperator"]:
+        """List of operators that provide inputs for this operator."""
+        assert hasattr(
+            self, "_input_dependencies"
+        ), "PhysicalOperator.__init__() was not called."
+        return self._input_dependencies
+
+    def get_stats(self) -> StatsDict:
+        """Return recorded execution stats for use with DatasetStats."""
+        raise NotImplementedError
+
+    def get_metrics(self) -> Dict[str, int]:
+        """Returns dict of metrics reported from this operator.
+
+        These should be instant values that can be queried at any time, e.g.,
+        obj_store_mem_allocated, obj_store_mem_freed.
+        """
+        return {}
+
+    def __reduce__(self):
+        raise ValueError("PhysicalOperator is not serializable.")
+
+    def __str__(self):
+        if self.input_dependencies:
+            out_str = ", ".join([str(x) for x in self.input_dependencies])
+            out_str += " -> "
+        else:
+            out_str = ""
+        out_str += f"{self.__class__.__name__}[{self._name}]"
+        return out_str
+
+    def num_outputs_total(self) -> Optional[int]:
+        """Returns the total number of output bundles of this operator, if known.
+
+        This is useful for reporting progress.
+        """
+        if len(self.input_dependencies) == 1:
+            return self.input_dependencies[0].num_outputs_total()
+        return None
+
+    def add_input(self, refs: RefBundle, input_index: int) -> None:
+        """Called when an upstream result is available.
+
+        Inputs may be added in any order, and calls to `add_input` may be interleaved
+        with calls to `get_next` / `has_next` to implement streaming execution.
+
+        Args:
+            refs: The ref bundle that should be added as input.
+            input_index: The index identifying the input dependency producing the
+                input. For most operators, this is always `0` since there is only
+                one upstream input operator.
+        """
+        raise NotImplementedError
```

Comment:
How do we plan to propagate whether we need to preserve order down to operators at execution time?

Right now, the `MapOperator` in the `BulkExecutor` PR has hardcoded order preservation into its execution state, without a convenient `PhysicalOperator` API to change that behavior at execution time (*not* instantiation time). It looks like we need a `PhysicalOperator` API for registering/binding execution options, maybe an API that's symmetrical to `PhysicalOperator.shutdown()`.

Reply:
Probably it will be added to execute()... maybe we should just omit it for now.